base/keylist/keylist.go

```diff
@@ -0,0 +1,218 @@
+	kl.initIndexes()
+	if idx, ok := kl.indexes[key]; ok {
+		kl.Values[idx] = val
+		kl.Keys[idx] = key
+		return
+	}
+	kl.indexes[key] = len(kl.Values)
+	kl.Values = append(kl.Values, val)
+	kl.Keys = append(kl.Keys, key)
+}
+
+// Insert inserts the given value with the given key at the given index.
+// This is relatively slow because it needs regenerate the keys list.
+// It returns an error if the key already exists because
+// the behavior is undefined in that situation.
+func (kl *List[K, V]) Insert(idx int, key K, val V) error {
+	if _, has := kl.indexes[key]; has {
+		return fmt.Errorf("keylist.Add: key %v is already on the list", key)
+	}
+
+	kl.Keys = slices.Insert(kl.Keys, idx, key)
+	kl.Values = slices.Insert(kl.Values, idx, val)
+	kl.newIndexes()
+	for i, k := range kl.Keys {
+		kl.indexes[k] = i
+	}
+	return nil
+}
+
+// ValueByKey returns the value corresponding to the given key,
+// with a zero value returned for a missing key. See [List.ValueByKeyTry]
+// for one that returns a bool for missing keys.
+func (kl *List[K, V]) ValueByKey(key K) V {
+	idx, ok := kl.indexes[key]
+	if ok {
+		return kl.Values[idx]
+	}
+	var zv V
+	return zv
+}
+
+// ValueByKeyTry returns the value corresponding to the given key,
+// with false returned for a missing key, in case the zero value
+// is not diagnostic.
+func (kl *List[K, V]) ValueByKeyTry(key K) (V, bool) {
+	idx, ok := kl.indexes[key]
+	if ok {
+		return kl.Values[idx], true
+	}
+	var zv V
+	return zv, false
+}
+
+// IndexIsValid returns an error if the given index is invalid.
+func (kl *List[K, V]) IndexIsValid(idx int) error {
+	if idx >= len(kl.Values) || idx < 0 {
+		return fmt.Errorf("keylist.List: IndexIsValid: index %d is out of range of a list of length %d", idx, len(kl.Values))
+	}
+	return nil
+}
+
+// IndexByKey returns the index of the given key, with a -1 for missing key.
+func (kl *List[K, V]) IndexByKey(key K) int {
+	idx, ok := kl.indexes[key]
+	if !ok {
+		return -1
+	}
+	return idx
+}
+
+// Len returns the number of items in the list.
+func (kl *List[K, V]) Len() int {
+	if kl == nil {
+		return 0
+	}
+	return len(kl.Values)
+}
+
+// DeleteIndex deletes item(s) within the index range [i:j].
+// This is relatively slow because it needs to regenerate the
+// index map.
+func (kl *List[K, V]) DeleteIndex(i, j int) {
+	ndel := j - i
+	if ndel <= 0 {
+		panic("index range is <= 0")
+	}
+	kl.Keys = slices.Delete(kl.Keys, i, j)
+	kl.Values = slices.Delete(kl.Values, i, j)
+	kl.newIndexes()
+	for i, k := range kl.Keys {
+		kl.indexes[k] = i
+	}
+
+}
+
+// DeleteKey deletes the item with the given key,
+// returning false if it does not find it.
+// This is relatively slow because it needs to regenerate the
+// index map.
+func (kl *List[K, V]) DeleteKey(key K) bool {
+	idx, ok := kl.indexes[key]
+	if !ok {
+		return false
+	}
+	kl.DeleteIndex(idx, idx+1)
+	return true
+}
+
+// Copy copies all of the entries from the given keyed list
```

Comment:
I specifically raised concerns that `keylist` would be confused with `KeyedList`, and you rejected them yet are making precisely the same mistake.